util/db/dbutil.go

```diff
@@ -56,32 +58,57 @@ type Accessor struct {
 	log      logging.Logger
 }
 
+func init() {
+	initCh = make(chan struct{}, 1)
+	initCh <- struct{}{}
+}
+
 // MakeAccessor creates a new Accessor.
 func MakeAccessor(dbfilename string, readOnly bool, inMemory bool) (Accessor, error) {
-	var db Accessor
-	db.readOnly = readOnly
-
-	var err error
-	db.Handle, err = sql.Open("sqlite3", URI(dbfilename, readOnly, inMemory)+"&_journal_mode=wal")
-
-	if err == nil {
-		err = db.runInitStatements()
-	}
-
-	return db, err
+	return makeAccessorImpl(dbfilename, readOnly, inMemory, []string{"_journal_mode=wal"})
 }
 
 // MakeErasableAccessor creates a new Accessor with the secure_delete pragma set;
 // see https://www.sqlite.org/pragma.html#pragma_secure_delete
 // It is not read-only and not in-memory (otherwise, erasability doesn't matter)
 func MakeErasableAccessor(dbfilename string) (Accessor, error) {
+	return makeAccessorImpl(dbfilename, false, false, []string{"_secure_delete=on"})
+}
+
+func makeAccessorImpl(dbfilename string, readOnly bool, inMemory bool, params []string) (Accessor, error) {
 	var db Accessor
-	db.readOnly = false
+	db.readOnly = readOnly
 
+	// SQLite3 driver we use (mattn/go-sqlite3) does not implement driver.DriverContext interface
+	// that forces sql.Open calling sql.OpenDB and return a struct without any touches to the underlying driver.
+	// Because of that SQLite library is not initialized until the very first call of sqlite3_open_v2 that happens
+	// in sql.DB.conn. SQLite initialization is not thread-safe on date of writing (2/27/2020) and
+	// mattn/go-sqlite3 has no special code to handle this case.
+	// Solution is to create a connection using a safe synchronization barrier right here.
+	// The connection goes to a connection pool inside Go's sql package and will be re-used when needed.
+	// See https://github.com/algorand/go-algorand/issues/846 for more details.
 	var err error
-	db.Handle, err = sql.Open("sqlite3", URI(dbfilename, false, false)+"&_secure_delete=on")
+	db.Handle, err = sql.Open("sqlite3", URI(dbfilename, readOnly, inMemory)+"&"+strings.Join(params, "&"))
 
 	if err == nil {
+		// create a connection to safely initialize SQLite
+		// sync logic: initCh contains only single value so only one thread reads it and obtains the connection
+		// it performs initialization and closes the channel so others are unblocked and go.
+		_, ok := <-initCh
```

Comment:
I think that sync.Once is a more elegant way to implement this.

Reply:
Done